crates/ruff_python_parser/src/string_token_flags.rs

```diff
@@ -0,0 +1,256 @@
+        Self::try_from(&value)
+    }
+}
+
+const DISALLOWED_USTRING_PREFIXES: StringFlagsInner = StringFlagsInner::B_PREFIX
+    .union(StringFlagsInner::F_PREFIX)
+    .union(StringFlagsInner::R_PREFIX);
+
+const DISALLOWED_BSTRING_PREFIXES: StringFlagsInner =
+    StringFlagsInner::U_PREFIX.union(StringFlagsInner::F_PREFIX);
+
+const DISALLOWED_FSTRING_PREFIXES: StringFlagsInner =
+    StringFlagsInner::U_PREFIX.union(StringFlagsInner::B_PREFIX);
+
+const DISALLOWED_RSTRING_PREFIXES: StringFlagsInner = StringFlagsInner::U_PREFIX;
+
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct StringFlags(StringFlagsInner);
+
+impl TryFrom<&char> for StringFlags {
+    type Error = String;
+
+    fn try_from(value: &char) -> Result<Self, String> {
+        Ok(Self(StringFlagsInner::try_from(value)?))
+    }
+}
+
+impl TryFrom<char> for StringFlags {
+    type Error = String;
+
+    fn try_from(value: char) -> Result<Self, String> {
+        Self::try_from(&value)
+    }
+}
+
+impl StringFlags {
+    pub const fn is_raw(self) -> bool {
+        self.0.contains(StringFlagsInner::R_PREFIX)
+    }
+
+    pub const fn is_bytestring(self) -> bool {
+        self.0.contains(StringFlagsInner::B_PREFIX)
+    }
+
+    pub const fn is_fstring(self) -> bool {
+        self.0.contains(StringFlagsInner::F_PREFIX)
+    }
+
+    pub const fn is_ustring(self) -> bool {
+        self.0.contains(StringFlagsInner::U_PREFIX)
+    }
+
+    pub const fn quote_style(self) -> QuoteStyle {
+        if self.0.contains(StringFlagsInner::DOUBLE) {
+            QuoteStyle::Double
+        } else {
+            QuoteStyle::Single
+        }
+    }
+
+    pub const fn is_triple_quoted(self) -> bool {
+        self.0.contains(StringFlagsInner::TRIPLE_QUOTED)
+    }
+
+    pub const fn quote_str(self) -> &'static str {
+        if self.is_triple_quoted() {
+            match self.quote_style() {
+                QuoteStyle::Single => "'''",
+                QuoteStyle::Double => r#"""""#,
+            }
+        } else {
+            match self.quote_style() {
+                QuoteStyle::Single => "'",
+                QuoteStyle::Double => "\"",
+            }
+        }
+    }
+
+    pub const fn prefix_str(self) -> &'static str {
+        if self.is_ustring() {
+            return "u";
+        }
+        if self.is_bytestring() {
+            if self.is_raw() {
+                return "rb";
+            }
+            return "b";
+        }
+        if self.is_fstring() {
+            if self.is_raw() {
+                return "rf";
+            }
+            return "f";
+        }
+        if self.is_raw() {
+            return "r";
+        }
+        ""
+    }
+
+    pub fn prefix_len(self) -> TextSize {
+        if self.is_ustring() {
+            return TextSize::from(1);
+        }
+        let len = u32::from(self.is_bytestring())
+            + u32::from(self.is_raw())
+            + u32::from(self.is_fstring());
+        len.into()
```

Comment:
Could perhaps just be `self. prefix_len().len()`?

Reply:
Something like this?

```rs
    pub fn prefix_len(self) -> TextSize {
        let len: u32 = self.prefix_str().len().try_into().unwrap();
        TextSize::from(len)
    }
```

Reply:
I thought there was `self.prefix_str().text_len()`, does that work?

Reply:
Didn't know about that -- that does work, thanks!